File: swift/Sources/GrpcDemo/Data/CatalogRepository.swift
// swift/Sources/GrpcDemo/Data/CatalogRepository.swift
//
// Implements CatalogRepositoryProtocol.
// Proto ↔ domain mapping is entirely contained here.
//
// The `Catalog_` prefix on generated types comes from protoc-gen-swift
// using the proto package name ("catalog") as a namespace.

import Foundation
import GRPC
import NIOCore

final class CatalogRepository: CatalogRepositoryProtocol {

    private let client: Catalog_CatalogServiceAsyncClient

    init(client: Catalog_CatalogServiceAsyncClient) {
        self.client = client
    }

    // ── Unary ─────────────────────────────────────────────
    func getProduct(id: String) async throws -> Product {
        do {
            let request = Catalog_GetProductRequest.with { $0.productID = id }
            let response = try await client.getProduct(request)
            return response.toDomain()
        } catch let error as GRPCStatus {
            throw error.toCatalogError()
        }
    }

    // ── Server-streaming ──────────────────────────────────
    // Swift's AsyncThrowingStream bridges gRPC streaming responses to
    // async/await — the same pattern as Kotlin's Flow<T>.
    func listProducts(category: String = "", page: Int = 1) -> AsyncThrowingStream<Product, Error> {
        AsyncThrowingStream { continuation in
            Task {
                do {
                    let request = Catalog_ListProductsRequest.with {
                        $0.category = category
                        $0.page     = Int32(page)
                        $0.pageSize = 20
                    }
                    for try await proto in client.listProducts(request) {
                        continuation.yield(proto.toDomain())
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }

    func watchPrice(productId: String) -> AsyncThrowingStream<PriceUpdate, Error> {
        AsyncThrowingStream { continuation in
            Task {
                do {
                    let request = Catalog_WatchPriceRequest.with { $0.productID = productId }
                    for try await proto in client.watchPrice(request) {
                        continuation.yield(proto.toDomain())
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }
}

// ── Proto → domain mappers ─────────────────────────────────
private extension Catalog_Product {
    func toDomain() -> Product {
        Product(
            id:         id,
            name:       name,
            pricePaise: pricePaise,
            imageURL:   URL(string: imageURL),
            inStock:    inStock,
            category:   category
        )
    }
}

private extension Catalog_PriceUpdate {
    func toDomain() -> PriceUpdate {
        PriceUpdate(productId: productID, pricePaise: pricePaise, currency: currency)
    }
}

// ── gRPC status → domain errors ────────────────────────────
private extension GRPCStatus {
    func toCatalogError() -> CatalogError {
        switch code {
        case .notFound:    return .notFound(id: message ?? "?")
        case .unavailable: return .unavailable(underlying: nil)
        default:           return .unknown(underlying: nil)
        }
    }
}

File: android/app/src/main/kotlin/com/demo/grpcdemo/data/CatalogRepository.kt
package com.demo.grpcdemo.data

import com.demo.grpc.generated.PriceUpdate as ProtoPriceUpdate
import com.demo.grpc.generated.Product as ProtoProduct
import com.demo.grpcdemo.domain.exception.CatalogException
import com.demo.grpcdemo.domain.model.PriceUpdate
import com.demo.grpcdemo.domain.model.Product
import com.demo.grpcdemo.domain.repository.CatalogRepository
import io.grpc.Status
import io.grpc.StatusRuntimeException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext
import javax.inject.Inject

// Implements the domain CatalogRepository interface.
// Owns all proto↔domain mapping so proto types never cross this boundary.
class CatalogRepositoryImpl @Inject constructor(
    private val dataSource: GrpcCatalogDataSource,
) : CatalogRepository {

    override suspend fun getProduct(productId: String): Result<Product> =
        withContext(Dispatchers.IO) {
            runCatching {
                dataSource.getProduct(productId).toDomain()
            }.mapGrpcError()
        }

    override fun listProducts(category: String, page: Int): Flow<Product> =
        dataSource.listProducts(category, page, pageSize = 20)
            .map { it.toDomain() }
            .catch { e -> throw mapGrpcException(e) }

    override fun watchPrice(productId: String): Flow<PriceUpdate> =
        dataSource.watchPrice(productId)
            .map { it.toDomain() }
            .catch { e -> throw mapGrpcException(e) }
}

// ── Proto → Domain mappers ────────────────────────────────────
private fun ProtoProduct.toDomain() = Product(
    id         = id,
    name       = name,
    pricePaise = pricePaise,
    imageUrl   = imageUrl,
    inStock    = inStock,
    category   = category,
)

private fun ProtoPriceUpdate.toDomain() = PriceUpdate(
    productId  = productId,
    pricePaise = pricePaise,
    currency   = currency,
)

// ── gRPC status → domain exception mapping ────────────────────
private fun mapGrpcException(e: Throwable): CatalogException =
    if (e is StatusRuntimeException) {
        when (e.status.code) {
            Status.Code.NOT_FOUND       -> CatalogException.NotFound(e.status.description ?: "?")
            Status.Code.UNAVAILABLE     -> CatalogException.Unavailable(e)
            Status.Code.UNAUTHENTICATED -> CatalogException.Unauthenticated()
            else                        -> CatalogException.Unknown(e)
        }
    } else CatalogException.Unknown(e)

private fun <T> Result<T>.mapGrpcError(): Result<T> =
    onFailure { throw mapGrpcException(it) }
